File: submissions/0387-first-unique-character-in-a-string/solution.java
import java.util.*;
import java.util.stream.Collectors;

public class Solution {
    public int firstUniqChar(String s) {
        if (s == null) {
            return -1;
        } else {
            int strLen = s.length();
            if (strLen == 0) {
                return -1;
            } else {
                return fastSolution(s);
            }
        }
    }

    private int fastSolution(String s) {
        int strLen = s.length();
        Map<Character, Integer> firstShown = new HashMap<>();
        List<Integer> candidates = new ArrayList<>();
        Set<Character> shown = new HashSet<>();
        for (int i = 0; i < strLen; i++) {
            char currentChar = s.charAt(i);
            if (shown.contains(currentChar)) {
                Integer first = firstShown.get(currentChar);
                if (first != null) {
                    int index = candidates.indexOf(first);
                    candidates.remove(index);
                    firstShown.remove(currentChar);
                }
            } else {
                shown.add(currentChar);
                firstShown.put(currentChar, i);
                candidates.add(i);
            }
        }

        if (candidates.size() != 0) {
            return candidates.get(0);
        } else {
            return -1;
        }
    }
}

File: submissions/0387-first-unique-character-in-a-string/solution.py
class Solution:
    def firstUniqChar(self, s: str) -> int:
        if s == "":
            return -1
        shown = set()
        unique_chars = set()
        for i in range(len(s)):
            c = s[i]
            if c not in shown:
                shown.add(c)
                unique_chars.add(c)
            else:
                if c in unique_chars:
                    unique_chars.remove(c)

        if unique_chars:            
            for i in range(len(s)):
                c = s[i]
                if c in unique_chars:
                    return i
        return -1


        
